pkg/apis/clusterapi/validation/cluster_test.go

```diff
@@ -0,0 +1,153 @@
+)
+
+func getValidClusterProviderConfig() *clusteroperator.ClusterProviderConfigSpec {
+	spec := &clusteroperator.ClusterProviderConfigSpec{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ClusterProviderConfigSpec",
+			APIVersion: "clusteroperator.openshift.io/v1alpha1",
+		},
+		ClusterSpec: clusteroperator.ClusterSpec{
+			ClusterVersionRef: clusteroperator.ClusterVersionReference{
+				Name: "someclusterref",
+			},
+			MachineSets: []clusteroperator.ClusterMachineSet{
+				{
+					ShortName: "master",
+					MachineSetConfig: clusteroperator.MachineSetConfig{
+						NodeType: "Master",
+						Size:     1,
+					},
+				},
+				{
+					ShortName: "infra",
+					MachineSetConfig: clusteroperator.MachineSetConfig{
+						NodeType: "Compute",
+						Infra:    true,
+						Size:     1,
+					},
+				},
+			},
+		},
+	}
+	return spec
+}
+
+func getValidClusterWithEmptyProviderConfig() *cluster.Cluster {
+	cluster := &cluster.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "test-cluster",
+		},
+		Spec: cluster.ClusterSpec{
+			ProviderConfig: cluster.ProviderConfig{
+				Value: nil,
+			},
+		},
+	}
+	return cluster
+}
+
+func getRawExtensionFromProviderConfig(providerConfig *clusteroperator.ClusterProviderConfigSpec) *runtime.RawExtension {
+	serializer := jsonserializer.NewSerializer(jsonserializer.DefaultMetaFactory, clusteroperatorapi.Scheme, clusteroperatorapi.Scheme, false)
+	var buffer bytes.Buffer
+	serializer.Encode(providerConfig, &buffer)
+
+	return &runtime.RawExtension{
+		Raw: buffer.Bytes(),
+	}
+}
+
+func getValidCluster() *cluster.Cluster {
+	cluster := getValidClusterWithEmptyProviderConfig()
+	providerConfig := getValidClusterProviderConfig()
+
+	cluster.Spec.ProviderConfig.Value = getRawExtensionFromProviderConfig(providerConfig)
+
+	return cluster
+}
+
+func getValidClusterStrategy() *ClusterStrategy {
+	return &ClusterStrategy{}
+}
+
+// TestValidateCluster tests the Validate function.
+func TestValidateCluster(t *testing.T) {
+	cases := []struct {
+		name            string
+		clusterStrategy *ClusterStrategy
+		cluster         *cluster.Cluster
+		valid           bool
+	}{
+		{
+			name:            "valid",
+			clusterStrategy: getValidClusterStrategy(),
+			cluster:         getValidCluster(),
+			valid:           true,
+		},
+		{
+			name:            "invalid name",
+			clusterStrategy: getValidClusterStrategy(),
+			cluster: func() *cluster.Cluster {
+				c := getValidCluster()
+				c.Name = "###"
+				return c
+			}(),
+			valid: false,
+		},
+		{
+			name:            "invalid spec",
+			clusterStrategy: getValidClusterStrategy(),
+			cluster: func() *cluster.Cluster {
+				c := getValidClusterWithEmptyProviderConfig()
+				pc := getValidClusterProviderConfig()
+				pc.MachineSets[0].Size = 0
+				c.Spec.ProviderConfig.Value = getRawExtensionFromProviderConfig(pc)
+				return c
+			}(),
+			valid: false,
+		},
+	}
+
+	ctx := request.NewContext()
+	if ctx == nil {
+		t.Errorf("problem creating context")
```

Comment:
You want to use `t.Fatalf` here. Otherwise, if something in the tests try to use the context, the tests will fail with a panic.